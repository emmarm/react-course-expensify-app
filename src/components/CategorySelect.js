import React from 'react';
import { connect } from 'react-redux';

export const CategorySelect = (props) => (
  <div>
    <select
      className="select"
      value={props.value}
      onChange={(e) => {
        const category = e.target.value;
        props.onCategoryChange(category);
      }}
    >
      {props.filter &&
        <option value='all'>
          All Categories
        </option>
      }
      <option value='general'>
        General
      </option>
      <option value='food'>
        Food
      </option>
      <option value='household'>
        Household
      </option>
      <option value='transportation'>
        Transportation
      </option>
      <option value='utilities'>
        Utilities
      </option>
      <option value='personal'>
        Personal
      </option>
      <option value='social'>
        Social
      </option>
    </select>
  </div>
);

export default connect()(CategorySelect);