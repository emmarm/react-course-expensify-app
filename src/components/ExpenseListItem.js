import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';


export default ({ id, description, amount, createdAt, note }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>${(amount / 100).toFixed(2)}</p>
    <p>{moment(createdAt).format('MMM Do, YYYY')}</p>
    <p>{note}</p>
  </div>
);