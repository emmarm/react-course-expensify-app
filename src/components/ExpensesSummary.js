import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
  const numberOf = props.expenses.length;
  const s = props.expenses.length !== 1 ? 's' : '';
  const total = getExpensesTotal(props.expenses);
  const formattedTotal = numeral(total).format('$0,0.00');
  return (
    <div>
      <p>Viewing {numberOf} expense{s} totalling {formattedTotal}</p>
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);