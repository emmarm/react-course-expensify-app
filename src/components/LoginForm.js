import React from 'react';
import { startLoginWithEmail } from '../actions/auth';

const INITIAL_STATE = {
  email: '',
  password: '',
  error: null
};

class LoginForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = { ...INITIAL_STATE };
  }
  onSubmit = (event) => {
    const {
      email,
      password
    } = this.state;
    startLoginWithEmail(email, password)
      .then(() => {
        this.setState(() => ({ ...INITIAL_STATE }));
        props.history.push('/');
      })
      .catch(error => {
        this.setState(() => ({
          error
        }));
      });
    event.preventDefault();
  }
  onEmailChange = (event) => {
    const email = event.target.value;
    this.setState(() => ({
      email
    }));
  }
  onPasswordChange = (event) => {
    const password = event.target.value;
    this.setState(() => ({
      password
    }));
  }
  render() {
    const {
      email,
      password,
      error
    } = this.state;
    const isInvalid =
      password === '' ||
      email === '';
    return (
      <form onSubmit={this.onSubmit} className="form">
        <input
          value={email}
          onChange={this.onEmailChange}
          type="text"
          placeholder="Email Address"
          className="text-input"
        />
        <input
          value={password}
          onChange={this.onPasswordChange}
          type="password"
          placeholder="Password"
          className="text-input"
        />
        <button disabled={isInvalid} type="submit" className="button">
          Log in
        </button>
        {error && <p>{error.message}</p>}
      </form>
    );
  }
};

export default LoginForm;