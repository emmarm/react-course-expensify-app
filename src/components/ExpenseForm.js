import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { CategorySelect } from './CategorySelect';

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? parseFloat(props.expense.amount / 100).toFixed(2) : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      category: props.expense ? props.expense.category : 'general',
      note: props.expense ? props.expense.note : '',
      error: ''
    };
  }
  onSubmit = (e) => {
    e.preventDefault();

    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please include a description and amount.'
      }));
    } else {
      this.setState(() => ({
        error: ''
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        category: this.state.category,
        note: this.state.note
      });
    }
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    const toTwoDecimalPlaces = /^\d+(\.\d{0,2})?$/;
    if (!amount || amount.match(toTwoDecimalPlaces)) {
      this.setState(() => ({
        amount
      }));
    }
  };
  onAmountBlur = (e) => {
    const amountEntered = e.target.value;
    let amount;
    if (amountEntered) {
      amount = parseFloat(amountEntered).toFixed(2);
    } else {
      amount = '';
    }
    this.setState(() => ({ amount }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }));
  };
  onCategoryChange = (category) => {
    this.setState(() => ({ category }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };
  render () {
    return (
      <form 
        className="form"
        onSubmit={this.onSubmit}
      >
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type='text'
          className="text-input"
          placeholder='Description'
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type='text'
          className="text-input"
          placeholder='Amount'
          value={this.state.amount}
          onChange={this.onAmountChange}
          onBlur={this.onAmountBlur}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <CategorySelect
          value={this.state.category}
          onCategoryChange={this.onCategoryChange}
        />
        <textarea
          className="textarea"
          placeholder='Add a note about this expense (optional)'
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <div>
          <button className="button">
            {this.props.expense ? 'Save Expense' : 'Add Expense'}
          </button>
        </div>
        <div>
          <Link to="/dashboard" className="button button--secondary">
            Cancel
          </Link>
        </div>
      </form>
    )
  };
};