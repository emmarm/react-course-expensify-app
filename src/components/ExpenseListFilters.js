import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, setCategory } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import CategorySelect from './CategorySelect';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
    filtersHidden: true
  };
  toggleHidden = () => {
    this.setState((prevState) => ({
      filtersHidden: !prevState.filtersHidden
    }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({
      calendarFocused
    }));
  };
  onCategoryChange = (category) => {
    this.props.setCategory(category);
  };
  render() {
    return (
      <div className="input-wrapper">
        <button 
          className="button button--toggle"
          onClick={this.toggleHidden}
        >
          {this.state.filtersHidden === true ? 'Show filters ⇓' : 'Hide filters ⇑'}
        </button>
        <div className={this.state.filtersHidden === true ? 'hidden' : null}>
          <div className="input-group">
            <div className="input-group__item">
              <input
                type="text"
                className="text-input text-input--narrow"
                placeholder="Search in expenses"
                value={this.props.filters.text}
                onChange={this.onTextChange}
              />
            </div>
            <div className="input-group__row">
              <div className="input-group__item">
                <select
                  className="select"
                  value={this.props.filters.sortBy}
                  onChange={this.onSortChange}
                >
                  <option value='date'>
                    Date
                  </option>
                  <option value='amount'>
                    Amount
                  </option>
                </select>
              </div>
              <div className="input-group__item">
                <DateRangePicker
                  startDate={this.props.filters.startDate}
                  endDate={this.props.filters.endDate}
                  onDatesChange={this.onDatesChange}
                  focusedInput={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  showClearDates={true}
                  numberOfMonths={1}
                  isOutsideRange={() => false}
                />
              </div>
              <div className="input-group__item">
                <CategorySelect
                  value={this.props.filters.category}
                  onCategoryChange={this.onCategoryChange}
                  filter={true}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setCategory: (category) => dispatch(setCategory(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);