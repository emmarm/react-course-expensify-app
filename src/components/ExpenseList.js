import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpenseList = (props) => {
  const numberOfVisible = props.visibleExpenses.length;
  const s = props.visibleExpenses.length !== 1 ? 's' : '';
  const total = getExpensesTotal(props.visibleExpenses);
  const formattedTotal = numeral(total).format('$0,0.00');
  const numberOfHidden = props.expenses.length - numberOfVisible;
  return (
    <div className="content-container--narrow mobile-full-width">
      <div className="list-header">
        <p className="list-header__summary">Showing <span>{numberOfVisible}</span> expense{s} ({numberOfHidden} hidden)</p>
        <p className="list-header__summary">Total: <span>{formattedTotal}</span></p>
      </div>
      <div className="list-body">
        {props.visibleExpenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No expenses</span>
          </div>
        ) : (
          props.visibleExpenses.map((expense) =>
            <ExpenseListItem
              key={expense.id}
              {...expense}
            />
          )
        )}
      </div>
    </div>
  )
};

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    visibleExpenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);