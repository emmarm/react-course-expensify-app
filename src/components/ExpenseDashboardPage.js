import React from 'react';
import { Link } from 'react-router-dom';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <div className="page-header">
      <Link className="button button--wide" to="/create">
        Add Expense
      </Link>
    </div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;