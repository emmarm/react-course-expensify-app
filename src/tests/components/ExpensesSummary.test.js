import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('Should correctly render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should correctly render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});