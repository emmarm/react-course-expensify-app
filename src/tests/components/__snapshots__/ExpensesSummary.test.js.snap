// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     expense
     totalling 
    $1.00
  </p>
</div>
`;

exports[`Should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    4
     expense
    s
     totalling 
    $10.00
  </p>
</div>
`;
